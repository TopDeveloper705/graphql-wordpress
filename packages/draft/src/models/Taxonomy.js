import Model from './Model';
import { getUniqueSlug } from './utils';

export default class Taxonomy extends Model {
  constructor(context) {
    super(context);

    this.collection = context.db.collection('taxonomy');
  }

  all({ limit = 10, offset = 0, search = null }) {
    const criteria = {};
    if (search) {
      criteria.$text = { $search: search };
    }

    return this.collection
      .find(criteria)
      .sort({ name: 1 })
      .skip(offset)
      .limit(limit)
      .toArray();
  }

  async insert(doc) {
    const slug = await getUniqueSlug(this.collection, doc.name);
    const docToInsert = Object.assign({}, doc, {
      slug,
      createdAt: Date.now(),
      updatedAt: Date.now(),
    });
    const id = (await this.collection.insertOne(docToInsert)).insertedId;
    return id;
  }
}
