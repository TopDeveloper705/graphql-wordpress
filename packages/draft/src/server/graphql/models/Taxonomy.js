import Model from './Model';
import { getUniqueSlug } from './utils';

export default class Taxonomy extends Model {
  constructor(context) {
    super(context);

    this.collection = context.db.collection('taxonomy');
  }

  all() {
    return this.collection
      .find({})
      .sort({ name: 1 })
      .toArray();
  }

  async insert(doc) {
    const slug = await getUniqueSlug(this.collection, doc.name);
    const docToInsert = Object.assign({}, doc, {
      slug,
      createdAt: Date.now(),
      updatedAt: Date.now(),
    });
    const id = (await this.collection.insertOne(docToInsert)).insertedId;
    return id;
  }
}
