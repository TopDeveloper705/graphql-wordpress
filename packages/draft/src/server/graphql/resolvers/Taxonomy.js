import { parseConnection } from './utils/collection';

const resolvers = {
  Taxonomy: {
    id(taxonomy) {
      return taxonomy._id;
    },
  },
  Query: {
    taxonomies(root, args, { Taxonomy }) {
      return parseConnection(Taxonomy, { first: 100 });
    },

    taxonomy(root, { id, slug }, { Taxonomy }) {
      if (id) {
        return Taxonomy.findOneById(id);
      }
      return Taxonomy.findOneBySlug(slug);
    },
  },
  Mutation: {
    async createTaxonomy(root, { input }, { Taxonomy }) {
      const id = await Taxonomy.insert(input);
      return Taxonomy.findOneById(id);
    },

    async updateTaxonomy(root, { id, input }, { Taxonomy }) {
      await Taxonomy.updateById(id, input);
      return Taxonomy.findOneById(id);
    },

    removeTaxonomy(root, { id }, { Taxonomy }) {
      return Taxonomy.removeById(id);
    },
  },
};

export default resolvers;
