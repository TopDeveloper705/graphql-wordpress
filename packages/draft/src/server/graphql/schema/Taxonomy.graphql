type Taxonomy {
  id: ObjID!
  name: String
  plural: String
  slug: String
  description: String
}

type TaxonomyEdge {
  node: Taxonomy
  cursor: String
}

type TaxonomyConnection {
  count: Int
  edges: [TaxonomyEdge]
  pageInfo: PageInfo
}

input CreateTaxonomyInput {
  name: String!
  plural: String
  slug: String
  description: String
}

input UpdateTaxonomyInput {
  name: String
  plural: String
  slug: String
  description: String
}

extend type Query {
  taxonomies(
    first: Int
    after: String
    last: Int
    before: String
    search: String
  ): TaxonomyConnection
  taxonomy(id: ObjID, slug: String): Taxonomy
}

extend type Mutation {
  createTaxonomy(input: CreateTaxonomyInput!): Taxonomy
  updateTaxonomy(id: ObjID!, input: UpdateTaxonomyInput!): Taxonomy
  removeTaxonomy(id: ObjID!): Boolean
}
