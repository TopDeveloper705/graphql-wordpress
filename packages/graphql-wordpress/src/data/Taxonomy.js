import { toGlobalId } from 'graphql-relay';

class Taxonomy {
  getID() {
    return toGlobalId(this.constructor.name, this.slug);
  }

  static getEndpoint() {
    return process.env.WP_TAXONOMIES_ENDPOINT || 'graphql/v1/taxonomies';
  }
}

export default Taxonomy;
