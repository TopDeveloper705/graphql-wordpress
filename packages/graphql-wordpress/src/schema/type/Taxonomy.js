import { GraphQLObjectType, GraphQLList, GraphQLString, GraphQLBoolean } from 'graphql';
import { globalIdField } from 'field/identifier';
import description from 'field/description';
import RewriteType from 'type/Rewrite';
import LabelsType from 'type/Labels';

const TaxonomyType = new GraphQLObjectType({
  name: 'Taxonomy',
  description: 'A taxonomy type.',
  fields: {
    id: globalIdField(),
    name: {
      type: GraphQLString,
      description: 'Identifier for the object.',
      resolve: item => item.slug,
    },
    ...description,
    types: {
      type: new GraphQLList(GraphQLString),
      description: 'Types associated with the taxonomy.',
    },
    hierarchical: {
      type: GraphQLBoolean,
      description: 'Whether or not the taxonomy should have children.',
    },
    rewrite: {
      type: RewriteType,
      description: 'Information that can be used to create pretty permalinks.',
    },
    labels: {
      type: LabelsType,
    },
  },
});

export default TaxonomyType;
