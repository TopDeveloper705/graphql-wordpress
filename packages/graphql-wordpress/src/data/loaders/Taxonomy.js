import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import Taxonomy from 'data/Taxonomy';

// Dataloader expects IDs that can be read by the REST API

export default function getTaxonomyLoaders() {
  // there is no batch mechanism on this endpoint
  const endpoint = Taxonomy.getEndpoint();

  const taxonomyLoader = new Dataloader(slugs =>
    fetchData(endpoint).then(({ data: { body } }) => slugs.map(slug => body[slug]))
  );

  return {
    load: async slug => {
      const data = await taxonomyLoader.load(slug);
      return data ? Object.assign(new Taxonomy(), data) : null;
    },
  };
}
