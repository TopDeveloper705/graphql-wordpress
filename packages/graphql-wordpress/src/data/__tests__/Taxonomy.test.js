import Taxonomy from 'data/Taxonomy';

const backupEnv = Object.assign({}, process.env);

afterAll(() => {
  process.env = backupEnv;
});

describe('Test Taxonomy data access', () => {
  test('Get endpoint', () => {
    expect(Taxonomy.getEndpoint()).toMatchSnapshot();
  });

  test('Get backup endpoint', () => {
    delete process.env.WP_TAXONOMIES_ENDPOINT;
    expect(Taxonomy.getEndpoint()).toMatchSnapshot();
  });
});
