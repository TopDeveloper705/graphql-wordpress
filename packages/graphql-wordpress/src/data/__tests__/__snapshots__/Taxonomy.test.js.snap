// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Taxonomy data access Get backup endpoint 1`] = `"graphql/v1/taxonomies"`;

exports[`Test Taxonomy data access Get endpoint 1`] = `"graphql/v1/taxonomies"`;
